### DISCUSSION TITLE
how to pass a array of struct as parameter for ffi use case?

### DISCUSSION THREAD

**Discussion by guzuomuse** [vlang/v#17494] [Questions and Answers]
i want to build a shared dll use v.       after some tries. i can not handle it. below is the equivment of  **rust** code :
```

///////////////////////rust--code///////////////////////////

use std::slice;

#[repr(C, packed(1))]
pub struct MyStruct {
    pub x: f64,
    pub y: i32,
}
#[no_mangle]
pub extern "C" fn array_of_struct_as_parameter_test(
    my_struct_arr_ptr: *mut MyStruct,//this parameter is a reference of array of `MyStruct`
    my_struct_array_len: usize,
) {
    let mut my_struct_array = unsafe {
        let rs = slice::from_raw_parts_mut(my_struct_arr_ptr, my_struct_array_len);
        rs
    };
    // next, do something with `my_struct_array`
    for mut item in my_struct_array {
        (*item).x = 66.6;
        (*item).y = 88;
    }
}
///////////////////////rust--code///////////////////////////

```
below is the equivement of v code:
```
///////////////////////v--code///////////////////////////

[packed]

struct MyStruct {

	mut:

	x f64
	y int
}
[export: 'array_of_struct_as_parameter_test']
[direct_array_access]
fn array_of_struct_as_parameter_test(mut s_arr []MyStruct, arr_len int) {
	ptr := unsafe { s_arr.pointers() }
	for i in 0 .. arr_len {
		a_ptr := ptr[i]
		mut a := new_array_from_c_array_no_alloc(arr_len, arr_len, sizeof(MyStruct), a_ptr) as MyStruct
		a.x = 66.6
		a.y = 88
	}
}
///////////////////////v--code///////////////////////////
```

but v code  give the error "error: function `new_array_from_c_array_no_alloc` is private"

since the `built-in` function `new_array_from_c_array_no_alloc` can not be used as public . how could i do this just like rust does ? thanks

**Comment by MatejMagat305** [comment]
First of all `mut` is NOT `pub` and if I'm right, you can write atribute in one, so you should write:
```
[export: 'array_of_struct_as_parameter_test', direct_array_access] // I m not sure whether ',' or ';'
struct Mystruct{
pub mut:
    x f64
    y int
}
```

**Reply by guzuomuse** [reply]
@MatejMagat305 

thanks for you reply . but even make the variables with `pub` modifier still not work at all.
here is my code:


```
[packed]

pub struct MyStruct {

pub mut:
	x f64
	y int
}
[export: 'v_array_of_struct_as_parameter_test';direct_array_access]
fn array_of_struct_as_parameter_test(mut s_arr []MyStruct, arr_len u32)  {
	unsafe {
		for i in 0 .. arr_len {
			mut item := &s_arr[i]
			item.x=333.3
			item.y=88
			// vmemcpy(item,new_struct,1*sizeof(MyStruct))
		}
	}
}
```

**Reply by MatejMagat305** [reply]
what is error now?

**Reply by guzuomuse** [reply]
@MatejMagat305 



the code has no compile error. and indeed can compile to a `dll` file. but when call the function exported from the dll


we got the error:

`
Access violation at 0x00007FFF93E5A2AE write to 0x00007FF7DBBC9A9C in 'main.dll'
`

**Reply by guzuomuse** [reply]
compare wtih the rust code i wrote before. rust take the `s_arr` parameter with the form of `*mut MyStruct` which the same as when we just pass a **single** struct. so we must offer what 's the size of the array .that's why we need the second parmater `arr_len` and then we get the slice of array with the `slice::from_raw_parts_mut` function by given the ptr and len. then we can got the reference and do some update on it.

rust as long as compile success and then worked as expect.
the v function take's the `s_arr` with a form of array `[]MyStruct` that's a big difference
v compile success but sink into executing failure. the error access violation means we try to access the illegal memory address.
it's worth mentioning, when use 

i'v tried pass the `s_arr` parameter as rust does. but without success.

it's worth mentioning


when write code like this:


```
	new_struct:=MyStruct{

		x: 3.3

		y: 1111
	}
	vmemcpy(s_arr,new_struct,1*sizeof(MyStruct))


```
only the first element was update. that's an expected behavior. since the pointer of the first element

is equal to the pointer of the whole array. but i can not find a way to update  the other emenents.
